// viaden-modules/config/css-selector.js
(function () {

	"use strict";
	/*global */

	var selector;

	selector = {

		register: {

			registrationForm: '.registration-form', // div
			eMail: '#email', // text
			phoneNumber: '#cell', // number
			userName: '#userName', // text
			password: '#password', // text
			confirmPassword: '#confirm-password', // text
			title: '#title option:nth-child(2)', // select
			firstName: '#firstname', // text
			lastName: '#lastname', // text
			birthDate: '#birthdate', // text
			country: '#cor option[value=GB]', // select
			postcode: '#postcode', // text
			postcodeFind: '#postcodeButton', // button
			postcodeSuggestion: '#postcodeSuggestionSelect option[value="0"]', // select
			submit: 'input[type="submit"][name="submit"]', // button
			termsContent: '.accept-terms-popup .scroll-box.info-content', // div
			termsSubmit: '.accept-terms-popup [data-action="accept"]', // button
			deposit: '.deposit-btn' // button
		},

		joinNow: '.join-now-btn',
		topBarBackButton: '#topBarBackButton',

		login: {
			openLogin: '[href="/login"]',
			userName: '#username',
			password: '#password',
			loginSubmit: '.login-popup input[type="submit"]',
			loginSuccessPopup: '.login-success-popup'
		},

		modal: 'body .modal.overlay:last-child',
		modalClose: 'body .modal.overlay:last-child .alert-close',

		openFooter: '[href="#footer"]',
		closeFooter: '#contextAmount',
		footerLinks: ['.main-nav a[href="/promotions"]', '.main-nav a[href="/gaming/all"]', '.main-nav a[href="/"]'],
		home: '.main-nav a[href="/"]',

		logout: {
			openLogout: '#account a[href="/logout"]',
			logoutSubmit: '[href="/logout"]'
		},

		regVisa: {
			toDeposit: '[href="/payment/deposit"]',
			visaCard: '[href="/payment/addcard/VISA"]',
			visaCardIFrame: '#ThirdPartyPaymentIframe',
			cardNumber: '#cc_card_number',
			cardMonth: '#cc_exp_month option[value="11"]',
			cardYear: '#cc_exp_year option[value="23"]',
			submitCard: '#continueButton'
		},

		gameFinder: {
			wrapper: '.js-search-wrapper',
			gameFind: '.js-btn-games-search',
			searchField: '.js-search',
			resultList: '.games-finder-list',
			gamesCounter: '.js-games-counter',
			clearSearch: '.js-clear-search',
			backFade: '.js-close-search'
		}

	};


	module.exports = selector;

}());
