// viaden-modules/config/css-selector.js
(function () {

	"use strict";
	/*global */

	var selector;

	selector = {

		register: {

			registrationForm: '.registration-form', // div
			eMail: '#email', // text
			phoneNumber: '#cell', // number
			userName: '#userName', // text
			password: '#password', // text
			confirmPassword: '#confirm-password', // text
			title: '#title option:nth-child(2)', // select
			firstName: '#firstname', // text
			lastName: '#lastname', // text
			birthDate: '#birthdate', // text
			country: '#cor option[value=GB]', // select
			postcode: '#postcode', // text
			postcodeFind: '#postcodeButton', // button
			postcodeSuggestion: '#postcodeSuggestionSelect option[value="0"]', // select
			submit: 'input[type="submit"][name="submit"]', // button
			termsContent: '.accept-terms-popup .scroll-box.info-content', // div
			termsSubmit: '.accept-terms-popup [data-action="accept"]', // button
			deposit: '.deposit-btn' // button
		},

		joinNow: '.join-now-btn',

		login: {
			openLogin: '[href="/login"]',
			userName: '#username',
			password: '#password',
			loginSubmit: '.login-popup input[type="submit"]',
			loginSuccessPopup: '.login-success-popup'
		},

		modal: 'body .modal.overlay:last-child',
		modalClose: 'body .modal.overlay:last-child .alert-close',

		openFooter: '[href="#footer"]',

		logout: {

			openLogout: '#account a[href="/logout"]',
			logoutSubmit: '[href="/logout"]'

		},

		regVisa: {
			toDeposit: '[href="/payment/deposit"]',
			visaCard: '[href="/payment/addcard/VISA"]',
			visaCardIFrame: '#ThirdPartyPaymentIframe',
			cardNumber: '#cc_card_number',
			cardMonth: '#cc_exp_month option[value="11"]',
			cardYear: '#cc_exp_year option[value="23"]',
			submitCard: '#continueButton',
			invalidCardLabel: 'label[for="cc_card_number"]'
		}

	};



	module.exports = selector;

}());
