// viaden-modules/util/index.js
(function () {

	"use strict";
	/*global */

	var util, fs, webDriver, mainCfg, path, startPath;

	fs = require('fs');

	webDriver = require('selenium-webdriver');

	mainCfg = require('viaden-modules/config/main.js');

	path = require('path');

	startPath = process.cwd();

	util = {

		keyList: ['isDevice', 'isSendMail'],

		attributes: {

		},

		getRandomDigit: function (max) {
			return Math.floor( Math.random() * (max || 10) );
		},

		getCardNumber: function (numberLength) {

			numberLength = (numberLength || 16) - 1;

			var getRandomDigit = this.getRandomDigit,
				number;

			function getCardNumber() {

				var arr = [4],
					lastNumber,
					originalNumber,
					i;

				// Original Number
				for (i = 0; i < numberLength; i += 1) {
					arr.push( getRandomDigit() );
				}
				originalNumber = arr.join('');
				//console.log(arr);

				// Drop the last digit
				lastNumber = arr.pop();
				//console.log(arr);

				// Reverse the digits
				//arr.reverse();
				//console.log(arr);

				// Multiple odd digits by 2
				arr = arr.map(function (value, index) {
					var newValue = value * 2;
					if (newValue > 9) {
						newValue = newValue % 10 + 1;
					}
					return newValue;
				});

				//console.log(arr);

				// Subtract 9 to numbers over 9
				//arr = arr.map(function (value) {
				//	return value > 9 ? value - 9 : value;
				//});
				//console.log(arr);

				// Count control number
				var controlSum = 0;
				arr.forEach(function (value) {
					controlSum += value;
				});
				controlSum *= 9;

				controlSum = String(controlSum);

				var controlNumber = 10 - (parseInt( controlSum[controlSum.length - 1] ) || 10);

				return (controlNumber === lastNumber) && originalNumber;

			}

			do {
				number = getCardNumber();
			} while (!number);

			return number;

		},



		getCardNumberOld: function (numberLength) {

			numberLength = (numberLength || 13) - 2;

			var getRandomDigit = this.getRandomDigit,
				number;

			function getCardNumber() {

				var arr = [5, 2],
					lastNumber,
					originalNumber,
					i;

				// Original Number
				for (i = 0; i < numberLength; i += 1) {
					arr.push( getRandomDigit() );
				}
				originalNumber = arr.join('');
				//console.log(arr);

				// Drop the last digit
				lastNumber = arr.pop();
				//console.log(arr);

				// Reverse the digits
				arr.reverse();
				//console.log(arr);

				// Multiple odd digits by 2
				arr = arr.map(function (value, index) {
					return index % 2 ? value : value * 2;
				});
				//console.log(arr);

				// Subtract 9 to numbers over 9
				arr = arr.map(function (value) {
					return value > 9 ? value - 9 : value;
				});
				//console.log(arr);

				// Count control number
				var controlSum = 0;
				arr.forEach(function (value) {
					controlSum += value;
				});
				var controlNumber = controlSum % 10;

				//console.log(controlNumber);
				//console.log(lastNumber);

				return controlNumber === lastNumber && originalNumber;

			}

			do {
				number = getCardNumber();
			} while (!number);

			return number;

		},

		trueFn: function () {
			return true;
		},

		falseFn: function () {
			return false;
		},

		get: function (key) {
			return this.attributes[key];
		},

		set: function (key, value) {
			return this.attributes[key] = value;
		},

		createWebDriverClient: function(args) {

			args = args || {};

			var args = this.get('args'),
				driver = new webDriver
					.Builder()
					.usingServer(args.host)
					.withCapabilities({ browserName: args.browser })
					.build();

			driver.manage().timeouts().implicitlyWait(500);

			switch (args.size) {

				case 'mobile':

					// set mobile size

					break;

				case 'tablet':

					// set tablet size

					break;



			}

			//driver.manage().window().setSize(360, 540);


			return driver;

		},
		getTest: function (list) {

			if (list) {
				if (list.indexOf(',') !== -1) {
					return list.replace(/\s+/, '').replace(/,/g, '.js,').split(',');

				} else {
					return [ list.trim() + '.js' ];
				}
			}

			return fs.readdirSync( path.resolve(startPath, mainCfg.folder.test) );

		},
		getStartPath: function () {
			return startPath;
		},

		scrollTo: function(driver, selector) {
			return driver.executeScript("document.querySelector('" + selector + "').scrollIntoView(true); window.scrollBy(0, 100);");
		},

		argsInit: function () {

			var reIsBool = /^(true|false)$/,
				reIsNumber = /^((\d+(\.\d+)?)|(\.\d+))$/, // 1.2 || .3
				data = {},
				definedParams = mainCfg.definedParams,
				key;

			process.argv.forEach(function (val) {

				var arr, key, value;

				arr = val.split('=');
				key = arr.shift();
				value = arr.join('=');

				if ( key[0] === '-' ) {
					key = key.substr(1);
					data[key] = false;
					return;
				}

				if ( key[0] === '+' ) {
					key = key.substr(1);
					data[key] = true;
					return;
				}

				if ( reIsBool.test(value) ) {
					data[key] = 'true' === value;
					return;
				}

				if ( reIsNumber.test(value) ) {
					data[key] = parseFloat(value);
					return;
				}

				data[key] = value;

			}, this);

			if (data.host === true) { // detect extra string
				data.host = mainCfg.mobileHost;
			}

			for (key in definedParams) {
				if (definedParams.hasOwnProperty(key) && !data.hasOwnProperty(key)) {
					data[key] = definedParams[key];
				}
			}

			this.set('args', data);

			return data;

		}

	};

	util.argsInit();

	module.exports = util;

}());
